### PR TITLE
SI-9513 decrement "deleted" count in OpenHashMap.put() when slot reused

### PR DESCRIPTION
**Pull request by performantdata**

### CODE REVIEW

**Comment by performantdata** [comment]
review @Ichoran

**Comment by performantdata** [comment]
or @axel22 are you available for review?

File: src/library/scala/collection/mutable/OpenHashMap.scala [hunk 1]
```diff
@@ -81,6 +81,9 @@ extends AbstractMap[Key, Value]
     h ^ (h >>> 7) ^ (h >>> 4)
   }
 
+  /** Increase the size of the table.
+    * Copy only the occupied slots, effectively eliminating the deleted slots.
+    */
```
Comment (Ichoran) [inline src/library/scala/collection/mutable/OpenHashMap.scala:86]: As long as you are making documentation changes, can you make the return type on `growTable()` explicit?

File: test/junit/scala/collection/mutable/OpenHashMapTest.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+/** Tests for [[OpenHashMap]]. */
+@RunWith(classOf[JUnit4])
+class OpenHashMapTest {
+  /** Test that an [[OpenHashMap]] correctly maintains its internal `deleted` count. */
+  @Test
+  def maintainsDeletedCount {
+    import scala.reflect.runtime.{universe => ru}
```
Comment (Ichoran) [inline test/junit/scala/collection/mutable/OpenHashMapTest.scala]: This import isn't needed if you're using Java reflection.  If you want to write a Scala reflection version, can you put it in a second method that is entirely commented out, and just leave this one as a clean Java-reflection version?

**Comment by Ichoran** [comment]
Just minor fixes; the core logic seems fine.

File: src/library/scala/collection/mutable/OpenHashMap.scala (hunk 1 above)
Comment (performantdata) [inline src/library/scala/collection/mutable/OpenHashMap.scala:86]: One of my future PRs will touch `growTable()` code. I'll do it then. I only added the docs to help me parse the `put()` code.

File: test/junit/scala/collection/mutable/OpenHashMapTest.scala (hunk 2 above)
Comment (performantdata) [inline test/junit/scala/collection/mutable/OpenHashMapTest.scala]: I moved it into the comment. The commented code is there so to switch to Scala reflection once SI-9306 is fixed. I wouldn't call the Java version "clean", since, unlike Scala, it depends on the name being expanded. I'm just writing Java reflection code for other (future) tests in this class; will convert those as well after SI-9306.

**Comment by retronym** [comment]
Thanks for spotting this!

Could you please expand the commit comment of 1fb32fcf04386f52e72522bd688e69edafd95414 with some context for the change? I'm left wondering how the bug manifest itself: could it have resulted in incorrect answers from operations on a `Map`? Why does the test case need to peek into the internal state of the map rather than validate its behaviour?

**Comment by performantdata** [comment]
> Could you please expand the commit comment of 1fb32fc with some context for the change? I'm left wondering how the bug manifest itself

It's in the [bug report](https://issues.scala-lang.org/browse/SI-9513). Is that not the place for it?

> could it have resulted in incorrect answers from operations on a Map?

No.

> Why does the test case need to peek into the internal state of the map rather than validate its behaviour?

There is no behavior externalized. There is no `capacity()` method available, for instance. Most I could do is pull in a library like JOL and measure object memory usage, which isn't quite as accurate. And @lrytz [said this is kosher](https://groups.google.com/forum/#!topic/scala-internals/W_2UBjUtQUs).

**Comment by retronym** [comment]
Sorry, I hadn't read the bug. I tend to repeat that information (what was wrong / how it manifest / what I'm changing / why this change is the right one) in the commit comment. Having it in Git makes life a little easier when we're on archaeological digs through the history. No need to do that for this PR, but keep it in mind for your next contribution.

**Comment by performantdata** [comment]
I thought there was a way to make a reference like "SI-9513" in a GitHub commit message link back to the issue tracker. I've seen the other way in action, where JIRA hooks into the commits and links back to the VCS. But neither on this project.

Thinking about that more...we linked in-house instances of Subversion and JIRA, both ways.

**Comment by retronym** [comment]
I use a [user script](https://gist.github.com/retronym/8d09391032e80af9181d) to add links from the GitHub UI to the corresponding JIRA issue. We disabled the automated links from JIRA tickets to GitHub PRs as we didn't like the way that that JIRA plugin automatically closed all tickets that were mentioned in the commit comments.

Here's an recent example of how I tend to structure a commit comment: https://github.com/scala/scala/commit/f5dc96bb550a840ea150985125f52e025ac8ac49

But in any case this change:

LGTM

**Comment by SethTisue** [comment]
thank you @performantdata! much appreciated

File: src/library/scala/collection/mutable/OpenHashMap.scala [hunk 3]
```diff
@@ -92,8 +95,18 @@ extends AbstractMap[Key, Value]
     deleted = 0
   }
 
+  /** Return the index of the first slot in the hash table (in probe order)
+    * that either is empty, or is or was last occupied by the given key.
+    */
   private[this] def findIndex(key: Key) : Int = findIndex(key, hashOf(key))
 
+  /** Return the index of the first slot in the hash table (in probe order)
+    * that either is empty, or is or was last occupied by the given key.
```
Comment (axel22) [inline src/library/scala/collection/mutable/OpenHashMap.scala:104]: Judging from the `while` loop, the returned index will either be such that `table(index) == null` or `table(index).key == key`. To avoid confusion, I would write:

```
 *  that either is empty, or is occupied by an entry with the given key.
```

(the "was" part requires context to understand, so I suggest to remove it)

File: src/library/scala/collection/mutable/OpenHashMap.scala [hunk 4]
```diff
@@ -136,7 +149,11 @@ extends AbstractMap[Key, Value]
       None
     } else {
       val res = entry.value
```
Comment (axel22) [inline src/library/scala/collection/mutable/OpenHashMap.scala:151]: It's somewhat contentious whether or not the `entry.key` should also be updated at this point.

I would say yes, but do the change in a separate PR.

File: src/library/scala/collection/mutable/OpenHashMap.scala (hunk 1 above)
Comment (axel22) [inline src/library/scala/collection/mutable/OpenHashMap.scala:86]: I would also gently push for adding the explicit return type sooner rather than later, in case you make changes to this PR :)

Minor nitpick - we usually wrote docs like this in the past:

```
/** First line.
 *  Second line (star indented 1 whitespace).
 */
```

File: src/library/scala/collection/mutable/OpenHashMap.scala (hunk 4 above)
Comment (Ichoran) [inline src/library/scala/collection/mutable/OpenHashMap.scala:151]: That's not the kind of change you should make in a point release anyway.  You could have something that relies on the existing behavior that would unexpectedly stop working.

Comment (axel22) [inline src/library/scala/collection/mutable/OpenHashMap.scala:151]: Agreed - should be a PR to a different branch.

File: src/library/scala/collection/mutable/OpenHashMap.scala (hunk 3 above)
Comment (Ichoran) [inline src/library/scala/collection/mutable/OpenHashMap.scala:104]: No, "is or was" is correct and crucial to understanding what the routine does, which is to return the entry which might have `Some(value)` or might have `None` for the value meaning that it was deleted.  Hence "is or was".  Maybe it needs to be expanded to make that logic clear.

Comment (axel22) [inline src/library/scala/collection/mutable/OpenHashMap.scala:104]: Yes, that was my point - it would help to expand it, because understanding "or was" required scanning through the entire class.

Comment (performantdata) [inline src/library/scala/collection/mutable/OpenHashMap.scala:104]: Yes, there's undocumented semantics in the internal structure. I'm not trying to document everything (like it should have been when created), because I have other changes in mind that will remove said structure.

File: src/library/scala/collection/mutable/OpenHashMap.scala (hunk 1 above)
Comment (performantdata) [inline src/library/scala/collection/mutable/OpenHashMap.scala:86]: > in case you make changes to this PR

It's closed already. (But thanks for joining us!)

> Minor nitpick

That's what Scala IDE wants to do to my code, too, but I'm just following [the style guide](http://docs.scala-lang.org/style/scaladoc.html). And [the other style guide](https://wiki.scala-lang.org/display/SW/Syntax). Can't have too much style.

File: src/library/scala/collection/mutable/OpenHashMap.scala (hunk 4 above)
Comment (performantdata) [inline src/library/scala/collection/mutable/OpenHashMap.scala:151]: > the `entry.key` should also be updated at this point

That looks like it'd be practically a no-op. If `entry != null`, then `entry.key == key`. So maybe `entry.key ne key`, but why should this class care?

> It's somewhat contentious

Sounds like there's some history to this debate.

**Comment by performantdata** [comment]
@Ichoran @axel22 [SI-9520](https://issues.scala-lang.org/browse/SI-9520) may interest you too.

File: src/library/scala/collection/mutable/OpenHashMap.scala (hunk 4 above)
Comment (axel22) [inline src/library/scala/collection/mutable/OpenHashMap.scala:151]: It's a no-op only if key carries no extra information other than that implied by equality.

But, to rephrase what I said previously: I argue that in `remove`, the `key` should be nulled out, to avoid memory leaks in cases where `Key` is some expensive object (which is not the goal of this PR, but should probably be addressed separately).

I'm not aware if this was discussed previously on the mailing list.

Comment (performantdata) [inline src/library/scala/collection/mutable/OpenHashMap.scala:151]: > in `remove`, the `key` should be nulled out, to avoid memory leaks

OK. That's not how I understood your earlier comment. I haven't seen this discussed anywhere else yet. It also didn't occur to me, but I've been skeptical of that whole business of keeping the old `Entry`s around.

If it's contentious, people can hash it out :grin: on [SI-9522](https://issues.scala-lang.org/browse/SI-9522).
